<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>MetAMOS by marbl</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1 class="header">MetAMOS</h1>
        <p class="header">A metagenomic assembly analysis pipeline</p>

        <ul>
          <li class="download"><a class="buttons" href="https://github.com/marbl/metAMOS/zipball/master">Download ZIP</a></li>
          <li class="download"><a class="buttons" href="https://github.com/marbl/metAMOS/tarball/master">Download TAR</a></li>
          <li><a class="buttons github" href="https://github.com/marbl/metAMOS">View On GitHub</a></li>
        </ul>

        <p class="header">This project is maintained by <a class="header name" href="https://github.com/marbl">marbl</a></p>


      </header>
      <section>
        <h1>
<a name="metamos-v15rc1-praline-brownie-readme" class="anchor" href="#metamos-v15rc1-praline-brownie-readme"><span class="octicon octicon-link"></span></a>MetAMOS v1.5rc1 "Praline Brownie" README</h1>

<p>Last updated: February 3rd 2014</p>

<hr><p>We are happy to announce version 1.5rc1, a.k.a. Praline Brownie </p>

<h2>
<a name="table-of-contents" class="anchor" href="#table-of-contents"><span class="octicon octicon-link"></span></a>Table of Contents</h2>

<ul>
<li><a href="#news">NEWS</a></li>
<li><a href="#metamos-pyinstaller-single-file-binary">MetAMOS single file binary</a></li>
<li>
<a href="#summary">SUMMARY</a>

<ul>
<li>
<a href="#a-hardware-requirements">HARDWARE REQUIREMENTS</a>

<ul>
<li>32GB of RAM</li>
</ul>
</li>
<li>
<a href="#b-software-requirements">SOFTWARE REQUIREMENTS</a>

<ul>
<li>git</li>
<li>gcc</li>
<li>automake</li>
<li>python-tools</li>
<li>python-devel</li>
<li>zlib-devel</li>
<li>numpy</li>
<li>freetype, freetype-devel</li>
<li>libpng-devel</li>
<li>matplotlib</li>
<li>curl</li>
</ul>
</li>
<li>
<a href="#c-installing-metamos">INSTALLING METAMOS</a>

<ul>
<li>python INSTALL.py core</li>
</ul>
</li>
<li>
<a href="#d-quick-start">QUICK START</a>

<ul>
<li>initPipeline -q -1 fastq.1 -2 fastq.2 -d projectDir -W core</li>
<li>runPipeline -d projectDir -p 16</li>
</ul>
</li>
<li>
<a href="#e-workflows">WORKFLOWS</a>

<ul>
<li>core</li>
<li>imetamos</li>
</ul>
</li>
<li><a href="#f-generic-tools">GENERIC TOOL</a></li>
<li><a href="#g-test-suite">TEST SUITE</a></li>
<li><a href="#h-example-output">EXAMPLE OUTPUT</a></li>
<li><a href="#i-contact">CONTACT</a></li>
<li>
<a href="#j-cite">CITE</a>

<ul>
<li>PMID: 23320958</li>
</ul>
</li>
<li><a href="#k-issues">ISSUES</a></li>
</ul>
</li>
<li><a href="#end">FIN</a></li>
</ul><h2>
<a name="news" class="anchor" href="#news"><span class="octicon octicon-link"></span></a>NEWS</h2>

<pre><code>   1. iMetAMOS now available.
   2. Updated 64-bit frozen Linux binary now available (fixed FastQC issue)
   3. Generic framework for adding assemblers/classifiers
   4. SRA run identifiers supported.
   5. Remote input files, as well as compressed input files are supported.
   6. Numerous bug fixes
</code></pre>

<p>*on deck: kmher, and Viritas</p>

<hr><p><a href="#table-of-contents">TOC</a></p>

<h2>
<a name="metamos-pyinstaller-single-file-binary" class="anchor" href="#metamos-pyinstaller-single-file-binary"><span class="octicon octicon-link"></span></a>MetAMOS PyInstaller single file binary</h2>

<p>In attempt to further simplify the MetAMOS installation process, we are happy to announce the availability of a 'frozen' MetAMOS binary for Linux-x68_64 platforms. Along with this binary comes a significantly reduced list of prerequisites:</p>

<ul>
<li>Java 1.6 (or newer)</li>
<li>Perl 5.8.8 (or newer)</li>
<li>64-bit *nix OS or Mac OSX 10.7+ (you may need to install MacPorts for full functionality)</li>
</ul><p>Linux 64-bit: <a>ftp://ftp.cbcb.umd.edu/pub/data/treangen/MA_fb_v1.5rc1_linux.tar.gz</a></p>

<p>OSX 64-bit: <a>ftp://ftp.cbcb.umd.edu/pub/data/treangen/MA_fb_v1.5rc1_OSX.tar.gz</a></p>

<p>Blast DBS: <a>ftp://ftp.cbcb.umd.edu/pub/data/treangen//blastdbs.tar.gz</a></p>

<p>This binaries are BIG, not only to download but also to run. Please ensure you have ample free space (100GB+) before download &amp; use. Once you've downloaded &amp; installed the tarball, there is NO NEED to redownload the huge package, instead download the runPipeline updates from here:</p>

<p>Linux 64-bit: <a>ftp://ftp.cbcb.umd.edu/pub/data/treangen/runPipeline_fb.linux.tar.gz</a></p>

<p>OSX 64-bit: <a>ftp://ftp.cbcb.umd.edu/pub/data/treangen/runPipeline_fb.OSX.tar.gz</a></p>

<hr><p><a href="#table-of-contents">TOC</a></p>

<h2>
<a name="summary" class="anchor" href="#summary"><span class="octicon octicon-link"></span></a>SUMMARY</h2>

<pre><code>    A) HARDWARE REQUIREMENTS
    B) SOFTWARE REQUIREMENTS
    C) INSTALLING MetAMOS
    D) QUICK START
    E) WORKFLOWS
    F) GENERIC TOOL
    G) TEST SUITE
    H) EXAMPLE OUTPUT
    I) CONTACT
    J) CITE
    K) ISSUES
</code></pre>

<hr><p><a href="#table-of-contents">TOC</a></p>

<h3>
<a name="a-hardware-requirements" class="anchor" href="#a-hardware-requirements"><span class="octicon octicon-link"></span></a>A) HARDWARE REQUIREMENTS</h3>

<p>MetAMOS was designed to work on any standard 64bit Linux
environment. To use MetAMOS for tutorial/teaching purposes, a minimum
of 8 GB RAM is required. To get started on real data sets a minimum of
32 GB of RAM is recommended, and anywhere from 64-1000 GB may be
necessary for larger datasets. In our experience, for most 50-100
million read datasets, 64 GB is a good place to start (128 GB of memory now
available on High Memory Instance at Amazon Elastic Compute Cloud ). </p>

<hr><p><a href="#table-of-contents">TOC</a></p>

<h3>
<a name="b-software-requirements" class="anchor" href="#b-software-requirements"><span class="octicon octicon-link"></span></a>B) SOFTWARE REQUIREMENTS</h3>

<p>If you are using the frozen binary, you can skip this section. The MetAMOS frozen
binary includes dependencies and only requires perl 5.8+ and java 1.6+.</p>

<p>The main prerequisite software for installing/running MetAMOS is python 2.6+, perl 5.8+,
and java 1.6+. Depending on your platform/Linux distribution, you might also need to download and 
install the following BEFORE running INSTALL.py:</p>

<ol>
<li>git</li>
<li>gcc</li>
<li>automake</li>
<li>python-tools</li>
<li>python-devel</li>
<li>zlib-devel</li>
<li>numpy</li>
<li>freetype, freetype-devel</li>
<li>libpng-devel</li>
<li>matplotlib</li>
<li>curl</li>
<li>R</li>
</ol><p>Additional software will be downloaded by MetAMOS as needed.
Additionally, there is some software that MetAMOS can
incorporate into its pipeline that we are not allowed to distribute,
such as MetaGeneMark and Newbler. To get a license to use MetaGeneMark, please
visit: <a href="http://exon.gatech.edu/license_download.cgi">http://exon.gatech.edu/license_download.cgi</a>. Once the tool is installed,
add it to your PATH variable and MetAMOS will then enable its use in the pipeline.</p>

<hr><p><a href="#table-of-contents">TOC</a></p>

<h3>
<a name="c-installing-metamos" class="anchor" href="#c-installing-metamos"><span class="octicon octicon-link"></span></a>C) INSTALLING MetAMOS</h3>

<p>To download the software release package, go <a href="https://github.com/marbl/metAMOS/archive/v1.5rc1.zip">here</a>.
You can also browse the <a href="https://github.com/treangen/MetAMOS/tree/v1.5rc1">repository</a>
and click on Downloads. Once downloaded, simply unpack the files and
open the MetAMOS directory. Once inside the MetAMOS directory, run:</p>

<pre><code>python INSTALL.py
</code></pre>

<p>This will download and install the external dependencies which may 
take minutes or hours to download depending on your connection speed. 
metAMOS supports workflows to install subsets of tools for faster installation.
By default only the core dependencies are installed. To install iMetAMOS run</p>

<pre><code>python INSTALL.py iMetAMOS
</code></pre>

<p>You can run:</p>

<pre><code>python INSTALL.py -h
</code></pre>

<p>to get a listing of available workflows and programs. You can specify either
workflows or programs as arguments to INSTALL.py. For example, to install the 
core workflow plus PhyloSift, run</p>

<pre><code>python INSTALL.py core phylosift
</code></pre>

<p>To install the programs which are part of the optional workflow run</p>

<pre><code>python INSTALL.py optional
</code></pre>

<p>If all dependencies are downloaded (including optional/deprecated ones), this will take 
quite awhile to complete (dependending on internet connection &amp; server, anywhere from a few hours to 1-2 days).</p>

<hr><p><a href="#table-of-contents">TOC</a></p>

<h3>
<a name="d-quick-start" class="anchor" href="#d-quick-start"><span class="octicon octicon-link"></span></a>D) QUICK START</h3>

<p>Before you get started using MetAMOS/iMetAMOS a brief review of its design will
help clarify its intended use. MetAMOS gas two main components:</p>

<ol>
<li>initPipeline</li>
<li>runPipeline</li>
</ol><p>Below is a simple example of running of iMetAMOS to assemble an SRA dataset:</p>

<pre><code>initPipeline -q -1 SRR987657 -d projectDir -W iMetAMOS
runPipeline -d projectDir -p 16
</code></pre>

<p>The first component, initPipeline, is for creating new projects and
also initializing sequence libraries. Currently interleaved &amp;
non-interleaved fasta, fastq, and SFF files are supported. Input
files can be compressed (bzip2, gzip) and can reside on remote
servers (in this case the full URL must be specified). SRA run identifiers
are also supported. </p>

<p>The file-type flags (-f, -q, and -s) must be specified before the file.
Once specified, they remain in effect until a different file type is specified.</p>

<p>usage: initPipeline -f/-q -1 file.fastq.1 -2 file.fastq.2 -d projectDir -i 300:500 </p>

<p>options: -s -c -q, -f, -1, -2, -d, -m, -i</p>

<ul>
<li>-1: either non-paired file of reads or first file in pair, can be list of multiple separated by a comma</li>
<li>-2: second paired read file, can be list of multiple separated by a comma</li>
<li>-c:  fasta file containing contigs</li>
<li>-d: output project directory (required)</li>
<li>-f: boolean, reads are in fasta format (default is fastq)</li>
<li>-h: display help message</li>
<li>-i: insert size of library, can be list separated by commas for multiple libraries</li>
<li>-l: SFF linker type</li>
<li>-m: interleaved file of paired reads</li>
<li>-o: reads are in outtie orientation (default innie)</li>
<li>-q: boolean, reads are in fastq format (default is fastq)</li>
<li>-s/--sff: boolean, reads are in SFF format (default is fastq)</li>
<li>-W: string: workflow name (-W iMetAMOS will run iMetAMOS). A workflow can specify parameters as well as data. A workflow can be immutable in which case any command-line parameters will not be used. Otherwise, command-line parameters take priority over workflow defaults.</li>
</ul><p>For example, to input a:</p>

<p>(non-interleaved fastq, single library)</p>

<pre><code>initPipeline -q -1 file.fastq.1 -2 file.fastq.2 -d projectDir -i 300:500
</code></pre>

<p>(non-interleaved fasta, single library)</p>

<pre><code>initPipeline -f -1 file.fastq.1 -2 file.fastq.2 -d projectDir -i 300:500
</code></pre>

<p>(interleaved fastq, single library)</p>

<pre><code>initPipeline -q -m file.fastq.12  -d projectDir -i 300:500
</code></pre>

<p>(interleaved fastq, multiple libraries)</p>

<pre><code>initPipeline -q -m file.fastq.12,file2.fastq.12  -d projectDir -i 300:500,1000:2000
</code></pre>

<p>(interleaved fastq, multiple libraries, existing assembly)</p>

<pre><code>initPipeline -q -m file.fastq.12,file2.fastq.12 -c file.contig.fa -d projectDir -i 300:500,1000:2000
</code></pre>

<p>(non-interleaved remote fastq, single library)</p>

<pre><code>initPipeline -q -1 ftp://ftp.cbcb.umd.edu/pub/data/metamos/gage-b-rb.miseq.1.fastq.gz -2 ftp://ftp.cbcb.umd.edu/pub/data/metamos/gage-b-rb.miseq.2.fastq.gz -d projectDir -i 300:500
</code></pre>

<p>(unpaired SRA run using iMetAMOS)</p>

<pre><code>initPipeline 1 &lt;SRA RUN ID&gt; -d projectDir -W iMetAMOS
</code></pre>

<p>(paired-end SRA run using iMetAMOS)</p>

<pre><code>initPipeline -m &lt;SRA RUN ID&gt; -d projectDir -i 300:500 -W imetAMOS
</code></pre>

<p>The second component, runPipeline, takes a project directory as
input and runs the following steps by default:</p>

<ol>
<li>Preprocess</li>
<li>Assemble</li>
<li>FindORFs</li>
<li>Validate</li>
<li>FindRepeats</li>
<li>Abundance</li>
<li>Annotate</li>
<li>FunctionalAnnotation</li>
<li>Scaffold</li>
<li>Propagate </li>
<li>FindScaffoldORFs</li>
<li>Classify </li>
<li>Postprocess</li>
</ol><p>usage info:</p>

<p>usage: runPipeline [options] -d projectdir</p>

<ul>
<li>  -h = :   print help [this message]</li>
<li>  -j = :   just output all of the programs and citations then exit (default = NO)</li>
<li>  -v = :   verbose output? (default = NO)</li>
<li>  -d = : directory created by initPipeline (REQUIRED)</li>
</ul><p>[options]: [pipeline_opts] [misc_opts]</p>

<p>[pipeline_opts]: options that affect the pipeline execution</p>

<p>Pipeline consists of the following steps:</p>

<p>Preprocess, Assemble, FindORFS, MapReads, Validate, Abundance, Annotate,
  Scaffold, Propagate, Classify, Postprocess</p>

<p>Each of these steps can be referred to by the following options:</p>

<ul>
<li>  -f = : force this step to be run (default = NONE)</li>
<li>  -s = : start at this step in the pipeline (default = Preprocess)</li>
<li>  -e = : end at this step in the pipeline (default = Postprocess)</li>
<li>  -n = : step to skip in pipeline (default=NONE)</li>
</ul><p>For each step you can fine-tune the execution as follows</p>

<p>[Preprocess]</p>

<ul>
<li>  -t = :   enable filter of input reads (default = metAMOS, options = metAMOS, EA-UTILS, PBcR for PacBio sequences)</li>
<li>  -q = :   produce FastQC quality report for reads with quality information (fastq or sff)? (default = NO)</li>
</ul><p>[Assemble]</p>

<ul>
<li>  -a = : genome assembler to use (default = SOAPdenovo).
      This can also be a comma-separated list of assembler (for example: soap,velvet)
      in this case, all selected assemblers will be run and the best selected for subsequent analysis</li>
<li>  -k = : k-mer size to be used for assembly (default = auto-selected).</li>
<li>        This can also be a comma-separated list of kmers to use</li>
<li>  -o = :    min overlap length</li>
</ul><p>[MapReads]</p>

<ul>
<li>  -m = : read mapper to use? (default = bowtie)</li>
<li>  -i = :   save bowtie (i)ndex (default = NO)</li>
<li>  -b = :   create library specific per bp coverage of assembled contigs (default = NO)</li>
</ul><p>[FindORFS]</p>

<ul>
<li>  -g = : gene caller to use (default=FragGeneScan)</li>
<li>  -l = :    min contig length to use for ORF call (default = 300)</li>
<li>  -x = :    min contig coverage to use for ORF call (default = 3X)</li>
</ul><p>[Validate]</p>

<ul>
<li>  -X = : comma-separated list of validators to run on the assembly. (default = lap, supported = reapr,orf,lap,ale,quast,frcbam,freebayes,cgal,n50)</li>
<li>  -S = : comma-separated list of scores to use to select the winning assembly. By default, all validation tools specified by -X will be run. For each score, an optional weight can be specified as SCORE:WEIGHT. For example, LAP:1,CGAL:2 (supported = all,lap,ale,cgal,snp,frcbam,orf,reapr,n50)</li>
</ul><p>[Annotate]</p>

<ul>
<li>  -c = : classifier to use for annotation (default = FCP)</li>
<li>  -u = :   annotate unassembled reads (default = NO)</li>
</ul><p>[Classify]</p>

<ul>
<li>  -z = : taxonomic level to categorize at (default = class)</li>
</ul><p>[misc_opts]: Miscellaneous options</p>

<ul>
<li>  -r = :   retain the AMOS bank  (default = NO)</li>
<li>  -p = :    number of threads to use (be greedy!) (default=1)</li>
<li>  -4 = :   454 data (default = NO)</li>
</ul><p>For example, to enable read filtering:</p>

<pre><code>-t
</code></pre>

<p>and to enable meta-IDBA as the assembler:</p>

<pre><code>-a metaidba
</code></pre>

<p>And to use PhyloSift to annotate:</p>

<pre><code>-c phylosift
</code></pre>

<p>Any single step in the pipeline can be skipped by passing the
following parameter to runPipeline:</p>

<pre><code>-n,--skipsteps=Step1,..
</code></pre>

<p>MetAMOS reruns steps based on timestamp information, so if the input
files for a step in the pipeline hasn't changed since the last run, it
will be skipped automatically. However, you can forcefully run any step
in the pipeline by passing the following parameter to runPipeline:</p>

<pre><code>-f,--force=Step1,..
</code></pre>

<p>MetAMOS stores a summary of the input libraries in pipeline.ini 
in the working directory. The pipeline.conf file stores the list 
of programs available to MetAMOS. Finally, pipeline.run stores the 
selected parameters and programs for the current run. MetAMOS also stores 
detailed logs of all commands executed by the pipeline in Logs/COMMANDS.log 
and a log for each step of the pipeline in Logs/.log</p>

<p>Upon completion, all of the final results will be stored in the
Postprocess/out directory. A component, create_summary.py, takes
this directory as input and as output, generates an HTML page with
with summary statistics and a few plots. An optional component, create_plots.py,
takes one or multiple Postprocess/out directories as input and generates
comparative plots.</p>

<hr><p><a href="#table-of-contents">TOC</a></p>

<h3>
<a name="e-workflows" class="anchor" href="#e-workflows"><span class="octicon octicon-link"></span></a>E) WORKFLOWS</h3>

<p>A workflow is a text-file that specified command-line options and input sequences
required to run metAMOS. A workflow may optionally inherit options/data from other
workflows. A workflow may also be immutable if the parameters should not be modifiable
by a user. An example workflow is below:</p>

<pre><code>inherit:isolate
modify:True
command: -q -u -r -v -I -c kraken -p 16 -a spades,velvet-sc,abyss,ray,edena,sga,masurca,soap,soap2,velvet -t metamos -n FunctionalAnnotation -f Postprocess -z phylum
asmcontigs: /Users/skoren/Personal/Research/metAMOS/Test/test.asm,ftp://ftp.ncbi.nih.gov/genomes/Bacteria/Candidatus_Carsonella_ruddii_uid58773/NC_008512.fna
lib1format: fasta
lib1mated:  True
lib1innie:  True
lib1interleaved:    True
lib1f1: /Users/skoren/Personal/Research/metAMOS/Test/carsonella_pe_filt.fna.gz,2000,5000,3500,500
</code></pre>

<p>The available options are:</p>

<ul>
<li>  inherit - any other workflows to inherit from. In this case, the workflow inherits options from the isolate workflow</li>
<li>  modify  - whether users are allowed to specify command-line parameters at runtime. If false, command-line options are ignored</li>
<li>  command - command-line options to specify for runPipeline</li>
<li>  asmcontigs - optional, pre-assembled contigs to include in analysis. Can be remote file. Multiple files can be separated using commas.</li>
<li>  lib#format - input type for lib #. Can be fasta/fastq/sff</li>
<li>  lib#mated - whether the library is mated or not</li>
<li>  lib#innie - whether the mates are in the innie (Illumina paired-end) format or not (Illumina mate-pair)</li>
<li>  lib#interleaved - whether the input sequences are in a single file or in two separate files</li>
<li>  lib#f1  - the name of the input file, along with library min, max, mean, stdev</li>
</ul><p>An arbitrary number of libraries may be specified in the above format. The below example shows an unmated library:</p>

<pre><code>lib1format: fasta
lib1mated:  False
lib1innie:  False
lib1interleaved:    False
lib1frg:    /Users/skoren/Personal/Research/metAMOS/Test/carsonella_pe_filt.fna.gz
</code></pre>

<p>as well as a non-interleaved library:</p>

<pre><code>lib1format:     fasta
lib1mated:      True
lib1innie:      True
lib1interleaved:        False
lib1f1: /Users/skoren/Personal/Research/metAMOS/Test/carsonella_pe_1.fna.gz,2000,5000,3500,500
lib1f2: /Users/skoren/Personal/Research/metAMOS/Test/carsonella_pe.2.fna.gz,2000,5000,3500,500
</code></pre>

<p>Workflows may be shared between users, as long as the input files are accessible (i.e. they are on a remote server or the systems share a file system). Workflow files should be placed in the metAMOS/workflows directory or the working directory where MetAMOS is launched.</p>

<hr><p><a href="#table-of-contents">TOC</a></p>

<h3>
<a name="f-generic-tools" class="anchor" href="#f-generic-tools"><span class="octicon octicon-link"></span></a>F) GENERIC TOOLS</h3>

<p>MetAMOS allows new tools to be added to the ASSEMBLE and ANNOTATE steps without requiring code changes. The addition of a tool is a three-step process. </p>

<h6>
<a name="1-add-the-tool-name-under-metamosutilitiesstepnamegeneric-for-example-if-you-want-to-add-a-new-assembler-you-would-modify-assemblegeneric-this-file-contains-one-tool-name-per-line-the-tool-name-is-arbitrary-text-and-will-be-used-by-metamos-to-look-up-detailed-configuration-the-current-assemblegeneric-looks-like" class="anchor" href="#1-add-the-tool-name-under-metamosutilitiesstepnamegeneric-for-example-if-you-want-to-add-a-new-assembler-you-would-modify-assemblegeneric-this-file-contains-one-tool-name-per-line-the-tool-name-is-arbitrary-text-and-will-be-used-by-metamos-to-look-up-detailed-configuration-the-current-assemblegeneric-looks-like"><span class="octicon octicon-link"></span></a>1) Add the tool name under metAMOS/Utilities/&lt;STEPNAME&gt;.generic. For example. if you want to add a new assembler, you would modify ASSEMBLE.generic. This file contains one tool name per line. The tool name is arbitrary text and will be used by MetAMOS to look up detailed configuration. The current ASSEMBLE.generic looks like:</h6>

<pre><code>% cat Utilities/config/ASSEMBLE.generic
abyss
sga
spades
ray
masurca
mira
edena
idba-ud
</code></pre>

<p>You can add multiple versions of an assembler. In this documentation, we will add SOAPdenovo v1.05 in addition to the above tools. First, we will add soap_v105 to the end of ASSEMBLE.generic:</p>

<pre><code>% cat Utilities/config/ASSEMBLE.generic
abyss
sga
spades
ray
masurca
mira
edena
idba-ud
soap_v105
</code></pre>

<h6>
<a name="2-write-a-configuration-file-for-the-tool-the-configuration-file-specifies-input-requirements-for-the-program-as-well-as-a-name-output-and-executable-location-within-configuration-files-several-keywords-may-be-specified-that-are-updated-at-runtime-the-list-of-currently-supported-keywords-can-be-found-at-the-end-of-this-section-in-the-above-example-metamos-would-expect-a-file-named-soap_v105spec" class="anchor" href="#2-write-a-configuration-file-for-the-tool-the-configuration-file-specifies-input-requirements-for-the-program-as-well-as-a-name-output-and-executable-location-within-configuration-files-several-keywords-may-be-specified-that-are-updated-at-runtime-the-list-of-currently-supported-keywords-can-be-found-at-the-end-of-this-section-in-the-above-example-metamos-would-expect-a-file-named-soap_v105spec"><span class="octicon octicon-link"></span></a>2) Write a configuration file for the tool. The configuration file specifies input requirements for the program as well as a name, output, and executable location. Within configuration files, several keywords may be specified that are updated at runtime. The list of currently supported keywords can be found at the end of this section. In the above example, MetAMOS would expect a file named soap_v105.spec.</h6>

<p>Below is an example configuration file used for Ray:</p>

<pre><code>% cat Utilities/config/ray.spec 
[CONFIG]
maxlibs 1
input FASTQ
name Ray
output [PREFIX]_ray/Contigs.fasta
scaffoldOutput [PREFIX]_ray/Scaffolds.fasta
location cpp/[MACHINE]/Ray/bin
threads -n
paired_interleaved -i [FIRST]
paired -p [FIRST] [SECOND]
commands rm -rf [RUNDIR]/ray &amp;&amp; \
        [MPI] [THREADS] Ray \
        -o [RUNDIR]/[PREFIX]_ray [INPUT]
unpaired -s [FIRST]
[Ray]
k   [KMER]
</code></pre>

<p>The [CONFIG] section is the generic configuration section, you can specify step-specific configuration later on. Here, most properties of where the tool is located, what its output is, and what input it requires is specified:</p>

<ul>
<li>input - the type of input (FASTQ in this case)</li>
<li>name - the full name of the tool you want to report later on. This can be arbitrary text.</li>
<li>output - where the output contigs from the tool are. For assemblers, this is contigs. [PREFIX] is a keyword for the MetAMOS prefix for the assembly when it is run. This is assumed to be relative to the MetAMOS run directory.</li>
<li>scaffoldOutput - where the output scaffolds from the tool are, if available.</li>
<li>backupOutput - some assemblers fail to generate their final output on some datasets. In this case, this can specify preliminary contig output which will only be used if the main output is not available.</li>
<li>location - path to the executable. This is relative to metAMOS/Utilities. You can specify [MACHINE] to substitute your machine type into the executable path (i.e. Linux-x86_64). The user path will be searched if the tool is not found in the specified location</li>
<li>threads - the parameter to pass number of threads to use for the program, if available</li>
<li>paired - how to pass paired-end (assumed innie) interleaved data (FIRST refers to left mates, SECOND to right)</li>
<li>paired_interleaved - how to pass paired-end (assumed innie) non-interleaved. FIRST refers to the interleaved file.</li>
<li>mated - how to pass mate-pair data (assumed outtie) non-interleaved data (FIRST refers to left mates, SECOND to right)</li>
<li>mated_interleaved - how to pass mate-pair data (assumed outtie) interleaved mates</li>
<li>unpaired - how to pass fragment data to the program. FIRST refers to the unmated file.</li>
<li>commands - an arbitrary list of commands to run to execute the tool. Multiple lines are supported with the \ character. Multiple commands can be specified using &amp;&amp;. In the above example, rm -rf will run first followed by Ray. Common useful keywords are:

<ul>
<li>[PREFIX] - the prefix to use for output </li>
<li>[RUNDIR] where the program is running</li>
<li>[KMER] - the selected k-mer to use for assembly</li>
<li>[MEM] - available memory</li>
<li>[THREADS] - the threads parameter and number of threads requested by the user</li>
<li>[INPUT] - the formatted input based on the libraries provided to metAMOS</li>
</ul>
</li>
</ul><p>The [Ray] section is a step-specific configuration. This is based on the executable names used in commands above. By default the parameters will be passed with prefixed - so here Ray will be run with -k [KMER]</p>

<p>Some assemblers (SOAPdenovo, MaSuRCA, etc) require an input configuration file rather than taking parameters on the command line. In this case, we need both a spec and template file (soap_v105.spec and soap_v105.template) which will get updated at runtime and passed to the assembler. The [CONFIG] section then includes a config option which specifies the template and the keyword [INPUT] will pass the configuration file rather than library information. </p>

<p>Below is an example spec file for SOAPdenovo that requires a template and spec file:</p>

<pre><code>% cat Utilities/config/soap_v105.spec
[CONFIG]
input FASTQ
name soap_v105
threads -p
output [PREFIX]/[PREFIX].asm.contig
location cpp/[MACHINE]/SOAPdenovo_1.05/
scaffoldOutput [PREFIX]/[PREFIX].asm.scafSeq
config config/soap_v105.template
mated rank=[LIB]\navg_ins=[MEAN]\nreverse_seq=1\nasm_flags=2\nq1=[FIRST]\nq2=[SECOND]
paired rank=[LIB]\navg_ins=[MEAN]\nreverse_seq=0\nasm_flags=3\nq1=[FIRST]\nq2=[SECOND]
unpaired rank=[LIB]\navg_ins=0\nq=[FIRST]
commands rm -rf [PREFIX] &amp;&amp; \
         mkdir [PREFIX] &amp;&amp; \
     SOAPdenovo all -s [INPUT] -o [PREFIX]/[PREFIX].asm -K [KMER] [THREADS]
</code></pre>

<pre><code>% cat Utilities/config/soap_v105.template
#maximal read length
max_rd_len=150
[LIB]
[INPUT]
</code></pre>

<p>Here, the config template is specified (again relative to metAMOS/Utilities) and the [INPUT] keyword will be replaced by the library information at run time.</p>

<h6>
<a name="3-add-a-citation-to-the-tool-under-metamosdoccitationsrst" class="anchor" href="#3-add-a-citation-to-the-tool-under-metamosdoccitationsrst"><span class="octicon octicon-link"></span></a>3) Add a citation to the tool under metAMOS/doc/citations.rst</h6>

<p>Citations are tab-delimited and specify the lower-case tool alias, full tool-name, and citation information. For example:</p>

<pre><code>soap_v105   SOAPdenovo v1.05    Li Y, Hu Y, Bolund L, Wang J: State of the art de novo assembly of human genomes from massively parallel sequencing data.Human genomics 2010, 4:271-277.
</code></pre>

<p>The citation will be automatically printed by MetAMOS whenever a run uses the specified tool. </p>

<h6>
<a name="4-for-annotate-tools-we-also-need-a-way-to-convert-the-output-to-krona-by-default-metamos-will-look-for-an-importtoolnamepl-script-if-one-is-not-found-it-will-rely-on-a-generic-import-which-will-assumed-a-tab-delimited-format" class="anchor" href="#4-for-annotate-tools-we-also-need-a-way-to-convert-the-output-to-krona-by-default-metamos-will-look-for-an-importtoolnamepl-script-if-one-is-not-found-it-will-rely-on-a-generic-import-which-will-assumed-a-tab-delimited-format"><span class="octicon octicon-link"></span></a>4) For ANNOTATE tools, we also need a way to convert the output to Krona. By default, MetAMOS will look for an Import&lt;toolName&gt;.pl script. If one is not found, it will rely on a generic import which will assumed a tab-delimited format:</h6>

<pre><code>contig/readID   NCBI Taxonomy ID
</code></pre>

<p>The currently supported list of keywords:</p>

<ul>
<li>  MEM - max memory limit</li>
<li>  LIB - library identifier (i.e. 1, 2, 3, etc)</li>
<li>  INPUT - replace with input to the program (a collection of input files or libraries depending on the step or a configuration file)</li>
<li>  MACHINE - replaced with Linux-x86_64, Darwin-x86_64, etc</li>
<li>  FIRST - replaced with left mates in mated read or interleaved or unpaired reads otherwise</li>
<li>  SECOND - replaced with right mates, in paired non-interleaved libs</li>
<li>  ORIENTATION - replaced with the word innie or outtie</li>
<li>  ORIENTATION_FIGURE - replaced with ---&gt; &lt;--- or &lt;--- ---&gt; for pe and mp, respectively</li>
<li>  MEAN - replaced with library mean</li>
<li>  SD - replaced with library standard dev</li>
<li>  MIN - replaced with library min</li>
<li>  MAX - replaced with library max</li>
<li>  THREADS - replaced with thread parameter specified and requested number of threads</li>
<li>  KMER - the kmer requested</li>
<li>  OFFSET - the phred offset (33/64) of the input files</li>
<li>  PREFIX - the desired prefix for the program output</li>
<li>  DB - the location of the MetAMOS DBs (i.e. Utilities/DB)</li>
<li>  RUNDIR - the location where the program is running (i.e. MetAMOS run directory)</li>
<li>  LOCATION - the location where the program executable lives</li>
<li>  TECHNOLOGY - the type of sequencing data (454, Illumina, etc) </li>
</ul><hr><p><a href="#table-of-contents">TOC</a></p>

<h3>
<a name="g-test-suite" class="anchor" href="#g-test-suite"><span class="octicon octicon-link"></span></a>G) Test suite</h3>

<p>We have developed a set of scripts for testing the various features of
MetAMOS. All of these regression test scripts are available inside the
/Test directory and include all necessary datasets to run them. Here
is a brief listing of the test scripts we currently include:</p>

<p>*Test initPipeline
./Test/test_create.sh</p>

<p>*Vanilla test
./Test/run_test.sh</p>

<p>*Test PhlyoSift
./Test/test_amphora.sh</p>

<p>*Test Minimus
./Test/test_minimus.sh</p>

<p>*Test Preprocess filtration of non-interleaved fastq files
./Test/test_filter_noninterleaved_fastq.sh</p>

<p>*Test iMetAMOS
./Test/test_ima.sh</p>

<p>*Test SRA download
./Test/test_sra.sh</p>

<p>*Test Newbler (if available)
./Test/test_newbler.sh</p>

<p>*Test CA (fasta)
./Test/test_ca_fasta.sh</p>

<p>*Test CA (fastq)
./Test/test_ca.sh</p>

<p>*Test SOAPdenovo
./Test/test_soap.sh</p>

<p>*Test MetaVelvet
./Test/test_metavelvet.sh</p>

<p>*Test SparseAssembler
./Test/test_sparse.sh</p>

<p>*Test Velvet
./Test/test_velvet.sh</p>

<p>*Test FCP
./Test/test_fcp.sh</p>

<p>*Test Spades
./Test/test_spades.sh</p>

<p>*Test BLAST
./Test/test_blast.sh</p>

<hr><p><a href="#table-of-contents">TOC</a></p>

<h3>
<a name="h-example-output" class="anchor" href="#h-example-output"><span class="octicon octicon-link"></span></a>H) Example output</h3>

<p>MetAMOS generates an interactive web page once a run successfully completes:
<a href="http://www.cbcb.umd.edu/%7Esergek/imetamos/gageb/Postprocess/out/html/summary.html">http://www.cbcb.umd.edu/~sergek/imetamos/gageb/Postprocess/out/html/summary.html</a></p>

<p>This includes summary statistics and taxonomic information based on Krona [1].
The easiest way to interact with the results is through the web interface.
The Postprocess/out directory contains the results of the analysis. By default, 
metAMOS uses the prefix "proba" (Galician for test). Thus, files will have the name "proba".*.</p>

<ul>
<li>
<p>abundance.krona.html</p>

<p>Krona [1] plot of abundances using the tool selected for abundance (MetaPhyler [2] by default)</p>
</li>
<li>
<p>annotate.krona.html   </p>

<p>Krona [1] plot of abundances using the tool selected for classification (Kraken [3] by default)</p>
</li>
<li>
<p>asm.scores        </p>

<p>Validation scores for each assembly/kmer combination run. Header contains information on scores generated</p>
</li>
<li>
<p>best.asm      </p>

<p>The name of the assembly/kmer combination that was selected as the best</p>
</li>
<li>
<p>.classified </p>

<p>Subdirectory containing each level of the selected taxonomy (class by default) and the contigs/reads/orfs belonging to each</p>
</li>
<li>
<p>.original.annots    </p>

<p>Tab-delimited taxonomic level assignments for each contig/unassembled read. Class IDs correspond to NCBI taxonomy IDs.</p>
</li>
<li>
<p>.original.reads.annots  </p>

<p>Tab-delimited taxonomic level assignments as above, where contigs are replaced with their constituent sequences.</p>
</li>
<li>
<p>.propagated.annots      </p>

<p>Tab-delimited file as above after assembly graph-based propagation of assignments to contigs.</p>
</li>
<li>
<p>.propagated.reads.annots    </p>

<p>Tab-delimited file as above after propagation and having contigs replaced with their constituent reads.</p>
</li>
<li>
<p>html              </p>

<p>HTML output from the pipeline. summary.html contains an interactive results view.</p>
</li>
<li>
<p>proba.bnk         </p>

<p>AMOS bank format of the assembly that can be visualized using Hawkeye.</p>
</li>
<li>
<p>proba.classify.txt        </p>

<p>The raw output of the abundances using the tool selected for abundance estimations (MetaPhyler [2] by default)</p>
</li>
<li>
<p>proba.ctg.cnt         </p>

<p>The number of sequences mapped to each assembly contig  </p>
</li>
<li>
<p>proba.ctg.cvg         </p>

<p>The coverage of each assembly contig</p>
</li>
<li>
<p>proba.ctg.fa          </p>

<p>The assembled contigs</p>
</li>
<li>
<p>proba.hits            </p>

<p>The raw output of the contig/unassembled reads classifications using the selected tool (Kraken [3]) by default.</p>
</li>
<li>
<p>proba.lib1.contig.reads   </p>

<p>The per-library assignment of sequences to contigs</p>
</li>
<li>
<p>proba.lib1.unaligned.fasta    </p>

<p>The per-library unassembled sequences</p>
</li>
<li>
<p>proba.scf.fa          </p>

<p>The assembled scaffolds</p>
</li>
<li>
<p>proba.motifs.fa           </p>

<p>The motifs within scaffolds identified by Bambus 2</p>
</li>
<li>
<p>proba.orf.faa</p>

<p>The protein sequences of identified open reading frames (ORFs) in the assembly and unassembled reads</p>
</li>
<li>
<p>proba.orf.fna</p>

<p>The fasta sequences of identified open reading frames (ORFs) in the assembly and unassembled reads</p>
</li>
<li>
<p>proba.scf.orf.faa</p>

<p>The protein sequences of identified open reading frames (ORFs) in the scaffolds</p>
</li>
<li>
<p>proba.scf.orf.fna</p>

<p>The protein sequences of identified open reading frames (ORFs) in the scaffolds</p>
</li>
<li>
<p>ref.fasta         </p>

<p>The recruited reference genome used for validation (iMetAMOS only)</p>
</li>
<li>
<p>ref.name          </p>

<p>The name of the recruited reference genome (iMetAMOS only)</p>
</li>
</ul><p>Additional details for each step are available under /out. This includes the raw
output (as well as any intermediate files) of any tools run during that step. For example, 
Annotate/out/proba.prokka includes the full Prokka annotation output. 
Assemble/out/abyss*/ contains the intermediate files output by ABySS. Additionally, 
since MetAMOS stores all of its results in an AMOS bank, the assemblies 
can be visualized with Hawkeye.</p>

<p>[1] Ondov BD, Bergman NH, Phillippy AM.. Interactive
metagenomic visualization in a Web browser. BMC Bioinformatics. 2011
Sep 30;12:385.  PMID: 21961884</p>

<p>[2] Liu B, Gibbons T, Ghodsi M, Treangen T, Pop M. Accurate and fast estimation of taxonomic profiles from metagenomic shotgun sequences. BMC Genomics. 2011;12 Suppl 2:S4. Epub 2011 Jul 27.</p>

<p>[3] Wood DE, Salzberg SL. Rapid phylogenetic sequence classification through repeated exact alignment. In preparation.</p>

<hr><p><a href="#table-of-contents">TOC</a></p>

<h3>
<a name="i-contact" class="anchor" href="#i-contact"><span class="octicon octicon-link"></span></a>I) CONTACT</h3>

<p>If you encounter any problems/bugs, please check the known issues pages:
<a href="https://github.com/treangen/MetAMOS/issues?direction=desc&amp;sort=created&amp;state=open">https://github.com/treangen/MetAMOS/issues?direction=desc&amp;sort=created&amp;state=open</a>
to see if it has already been documented.</p>

<p>If not, please report the issue either using the contact information below or 
by submitting a new issue online. Please include information on your run,
any output produced by runPipeline, as well as the pipeline.* files and the 
Log/ file (if not too large).</p>

<p>Who to contact to report bugs, forward complaints, feature requests:</p>

<p>Todd Treangen: <a href="mailto:treangen@gmail.com">treangen@gmail.com</a>
Sergey Koren: <a href="mailto:sergek@umd.edu">sergek@umd.edu</a></p>

<hr><p><a href="#table-of-contents">TOC</a></p>

<h3>
<a name="j-cite" class="anchor" href="#j-cite"><span class="octicon octicon-link"></span></a>J) CITE</h3>

<p>Treangen TJ*, Koren S*, Sommer DD, Liu B, Astrovskaya I, Ondov B,
Darling AE, Phillippy AM, Pop M.  MetAMOS: a modular and open source
metagenomic assembly and analysis pipeline. Genome Biol. 2013 Jan
15;14(1):R2. PMID: 23320958.</p>

<p>url: <a href="http://genomebiology.com/content/pdf/gb-2013-14-1-r2.pdf">http://genomebiology.com/content/pdf/gb-2013-14-1-r2.pdf</a></p>

<p>*Indicates both authors contributed equally to this work</p>

<p><a href="#table-of-contents">TOC</a></p>

<h3>
<a name="k-issues" class="anchor" href="#k-issues"><span class="octicon octicon-link"></span></a>K) ISSUES</h3>

<p>Here is a <a href="https://github.com/treangen/metAMOS/issues?state=open">link</a> to known, open issues.</p>

<h2>
<a name="fin" class="anchor" href="#fin"><span class="octicon octicon-link"></span></a>FIN</h2>
      </section>
      <footer>
        <p><small>Hosted on <a href="http://pages.github.com">GitHub Pages</a> using the Dinky theme</small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
		          <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-42944922-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
